crates/autopilot/src/domain/competition/auction_mechanism/single_winner.rs

```diff
@@ -0,0 +1,225 @@
+            improvement
+                .checked_div(right.executed_sell.0.into())
+                .map(|v| U256::try_from(v).expect("improvement in buy fits in U256"))
+                .unwrap_or_default()
+        };
+
+        // Record best execution per order
+        let mut best_executions = HashMap::new();
+        for other in others {
+            for (uid, execution) in other.solution().orders() {
+                best_executions
+                    .entry(uid)
+                    .and_modify(|best_execution| {
+                        if !improvement_in_buy(execution, best_execution).is_zero() {
+                            *best_execution = *execution;
+                        }
+                    })
+                    .or_insert(*execution);
+            }
+        }
+
+        // Check if the solution contains an order whose execution in the
+        // solution is more than `fairness_threshold` worse than the
+        // order's best execution across all solutions
+        let unfair = solution
+            .solution()
+            .orders()
+            .iter()
+            .any(|(uid, current_execution)| {
+                let best_execution = best_executions.get(uid).expect("by construction above");
+                let improvement = improvement_in_buy(best_execution, current_execution);
+                if improvement.is_zero() {
+                    return false;
+                };
+                tracing::debug!(
+                    ?uid,
+                    ?improvement,
+                    ?best_execution,
+                    ?current_execution,
+                    "fairness check"
+                );
+                // Improvement is denominated in buy token, use buy price to normalize the
+                // difference into eth
+                let Some(order) = auction.orders.iter().find(|order| order.uid == *uid) else {
+                    // This can happen for jit orders
+                    tracing::debug!(?uid, "cannot ensure fairness, order not found in auction");
+                    return false;
+                };
+                let Some(buy_price) = auction.prices.get(&order.buy.token) else {
+                    tracing::warn!(
+                        ?order,
+                        "cannot ensure fairness, buy price not found in auction"
+                    );
+                    return false;
+                };
+                buy_price.in_eth(improvement.into()) > fairness_threshold
+            });
+        !unfair
+    }
+}
+
+impl AuctionMechanism for SingleWinnerAuctionMechanism {
+    fn filter_solutions(
+        &self,
+        auction: &Auction,
+        input_solutions: &[Participant<Unranked>],
+    ) -> Vec<Participant<Unranked>> {
+        // Shuffle so that sorting randomly splits ties.
+        let mut solutions = input_solutions.to_vec();
+        solutions.shuffle(&mut rand::thread_rng());
```

Comment:
I believe this should probably happen for any auction mechanism so it could be moved outside.